scripts/filter_expected_warnings.py

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env python
+"""Script to filter out old warnings from doxygen."""
+
+import sys
+import re
+
+
+class DoxygenWarning(object):
+    """Doxygen warning class."""
+
+    def __init__(self, firstline, filename, warning):
+        self.firstline = firstline
+        self.filename = filename
+        self.warning = warning
+        self.otherlines = []
+
+    def equals_ignoring_path_and_line_number(self, other):
+        """Return true if warnings have same filename and warning message."""
+        if self.filename != other.filename:
+            return False
+        if self.warning != other.warning:
+            return False
+        return self.otherlines == other.otherlines
+
+    def print_with_prefix(self, prefix):
+        print(prefix + self.firstline)
+        for line in self.otherlines:
+            print(prefix + '  ' + line)
+
+
+class WarningsList(object):
+    """List of Doxygen warnings."""
+
+    def __init__(self, all_lines):
+        """Create a list of the warnings in this file."""
+        self.warnings_list = []
+        current = None
+
+        warning_start_expr = re.compile(r'[^:]+/([^:/]+):\d+: (warning.*$)')
+
+        for line in all_lines:
+            if line.isspace():
+                continue
+
+            # Allow comments in the list of expected warnings.
+            if line.startswith('#'):
+                continue
+
+            matched = warning_start_expr.match(line)
+            if matched:
+                filename = matched.group(1)
+                warning = matched.group(2)
+                current = DoxygenWarning(line.strip(), filename, warning)
+                self.warnings_list.append(current)
+            elif line.startswith('  '):
+                current.otherlines.append(line.strip())
+            else:
+                # Assuming all warnings are of the form [path:line: warning:...]
+                # (and the warnings about too many nodes have been filtered out).
+                print('Error filtering warnings: Unexpected input format.')
+                print('  Input:' + line)
+
+    def contains(self, warning):
+        """Check if a similar warning is in this list."""
+        for other in self.warnings_list:
+            if warning.equals_ignoring_path_and_line_number(other):
+                return True
+        return False
+
+    def is_not_subset_of(self, superset, prefix):
```

Comment:
:pick: I don't really like the way this function has a big side effect (printing things out) which isn't indicated from the function name. I'd call it something like `print_elements_not_in_other_list`, and then the fact that it also returns if it found anything would not be unexpected, whereas you've name it after the return value and the fact it prints a lot of things is unexpected.

Reply:
That sounds like a good idea.  Hannes didn't like the big side effect either (and I agree too), but we were having trouble agreeing on another approach - if he's happy with your suggestion then I'll go for that.

Reply:
Done.